Update Project Cache docs to reflect new functionality (#9329)
dotnet/msbuild: documentation/specs/project-cache.md

In this example, we will only consider the following as inputs, and thus part of the fingerprint:
- The global properties of the build request (eg `Configuration=Debug`, `Platform=AnyCPU`)
- The content hash of the project file
- The content hash of files defined in specific items we know contribute to the build, like `<Compile>` and `<Content>`
- The fingerprint of referenced projects

Again, this is for illustrative purposes and a real implementation will want to use additional state for fingerprinting depending on the environment in which it runs and the correctness requirements.

It can make sense for a fingerprint to be a hash of its inputs, so effectively is a byte array which can be represented by a string.

At the beginning of the build, the plugin's `BeginBuildAsync` method will be called. As part of the `CacheContext`, the plugin is either given the graph or the entry points to the graph for which it can create a graph from. The plugin can use this graph to do some initial processing, like predicting various inputs which a project is likely to use. This information can then be stored to help construct a fingerprint for a build request later.

## Cache storage

> Any storage mechanism can be used as a cache implementation, for example [Azure Blob Storage](https://azure.microsoft.com/en-us/products/storage/blobs/), or even just the local filesystem. At least in this example the only real requirement is that it can be used effectively as a key-value store. In many cases it can be useful for content to be keyed by its hash, and for the metadata file to be keyed by the fingerprint. In particular when content is keyed by hash, it is effectively deduplicated across multiple copies of the same file, which is common in builds.

Reviewer: ```suggestion
Any storage mechanism can be used as a cache implementation, for example [Azure Blob Storage](https://azure.microsoft.com/products/storage/blobs/), or even just the local filesystem. At least in this example the only real requirement is that it can be used effectively as a key-value store. In many cases it can be useful for content to be keyed by its hash, and for the metadata file to be keyed by the fingerprint. In particular when content is keyed by hash, it is effectively deduplicated across multiple copies of the same file, which is common in builds.
```

> For illustration purposes, consider our cache implementation is based on a simple filesystem which a separate metadata and content directory inside it. Under the metadata dir, each file is a metadata file where the filename matches the fingerprint it's describing. Under the content dir, each file is a content file where the filename matches the hash of the content itself.

Reviewer: ```suggestion
For illustration purposes, consider our cache implementation is based on a simple filesystem with a separate metadata and content directory inside it. Under the metadata dir, each file is a metadata file where the filename matches the fingerprint it's describing. Under the content dir, each file is a content file where the filename matches the hash of the content itself.
```

Once MSBuild is done building the project, it will call the plugin's `HandleProjectFinishedAsync`. Now the plugin knows the project is done and can process the results and add them to a cache. In general it's only useful to cache successful results, so the plugin should filter out non-success results. The `FileAccessContext` provided can then be used to retrieve the list of `FileAccessData` the plugin recieved. These `FileAccessData` can be processed to understand which files were read and written as part of the build.

In our example, we can use the read files to construct a fingerprint for the build request. We can then add the files written during the build ("outputs") to some cache implementation.

The plugin would then create some metadata describing the outputs (eg. the paths and hashes) and the serialized `BuildResult`, and associate it with the fingerprint and put that assocation in the cache.

> To illustrate this, consider a project with fingerprint `F` which wrote a single file `O` with hash `H` and had `BuildResult R`. The plugin could create a metadata file `M` which describes the outputs of the build (the path and hash of `O`) as well as the serialized `R`. Using the cache implementation described above, the plugin would write the following two files to the cache:
>  - `metadata/F -> M`
>  - `content/H -> O`

Reviewer: I found this kinda hard to follow, I was expecting R to appear in one of the files. It's not _wrong_ now that I read it a few times but if you have an easy rephrase I'd appreciate it. Maybe something like
```suggestion
To illustrate this, consider a project with fingerprint `F` which wrote a single file `O` with hash `H` and had `BuildResult R`. The plugin could create a metadata file `M` which describes the outputs of the build (the path and hash of `O`) as well as the serialized `R`. Using the cache implementation described above, the plugin would write the following two files to the cache:
 - `metadata/F -> M:'{outputs: {path: 'path/to/O', hash: H}, result: R}`
 - `content/H -> O`
```

- `BuildManager.EndBuild` calls `ProjectCacheBase.EndBuildAsync` on all discovered plugins. This allows plugins to perform any required cleanup work. This is a blocking call which will be awaited before th build can complete.
- Only the user provided top level build requests are checked against the cache. The build requests issued recursively from the top level requests are not checked against the cache, since it is assumed that users issue build requests in reverse toposort order. Therefore when a project builds its references, those references should have already been built and present in MSBuild's internal cache, provided either by the project cache plugin or real builds. In other words, projects which are not well-described in the graph (eg using `<MSBuild>` tasks directly) will not benefit from the cache.
- The plugin instance will get called in reverse topo sort order (from dependencies up towards dependents). This happens based on the requirement of using `/graph`. Building in reverse topo sort order is common between Visual Studio solution builds and higher build engines.

> - Plugins can function with and without a static graph. When a static graph is not provided, hints about the graph entry points are provided (details in Defining the "graph" when static graph is not available).

Author: Reworked some of these points.

    - Best: A real `BuildResult` from a previous build is provided. This can either be done by serializing the `HandleProjectFinishedAsync`, or when the plugin's infrastructure (e.g. CloudBuild or AnyBuild builder nodes) runs and caches the build, it can tell MSBuild to serialize the BuildResult to a file via [BuildParameters.OutputResultsCacheFile](/src/Build/BackEnd/BuildManager/BuildParameters.cs#L767) or the `/outputResultsCache` command line argument. Then, on cache hits, the plugins deserialize the `BuildResult` and send it back to MSBuild. This is the most correct option, as it requires neither guessing nor proxy targets. Whatever a previous build did, that's exactly what's returned.
      - Potential Issue: serialization format may change between writing and reading the `BuildResult`, especially if binary serialization is used.

## Cache add scenario

- Upon a cache miss, MSBuild will generally handle a request as normal, ie by building it.
- To facilitate the plugin being able to hande future builds, MSBuild can report file accesses as well as the build result for the BuildRequestData.

> - MSBuild disables the in-proc node and uses [Detours](https://github.com/microsoft/Detours) to observe file accesses of the worker nodes. It forwards this information to the plugin for it to use as desired.

Reviewer 2: Should there be a statement/section describing the performance cost of Detours/cache add overall?

>   - This functionality has some implementation restrictions so will require additional opt-in. Specifically, the `/ReportFileAccesses` command-line flag or by setting `BuildParameters.ReportFileAccesses` for programatic use of `BuildManager`. If this is not set, no file accesses will be reported to the plugin, however the plugin will still be notified of the build result.

Reviewer 3: Those 3 points seem a bit redundant, while each telling  something slightly different ('MSBuild can report file accesses', 'MSBuild forwards file accesses info to the plugin', 'Unless opted in, no file accesses are reported') - wouldn't it be better to mention it just once, while mentioning the opt-in mechanics?

Author: Reworked some of the verbiage

- As described above, it is recommended to serialize the `BuildResult` from `HandleProjectFinishedAsync` for later replay.

# APIs and calling patterns

## Plugin API
[ProjectCachePluginBase](/src/Build/BackEnd/Components/ProjectCache/ProjectCachePluginBase.cs) is an abstract class which plugin implementors will subclass. It contains the following methods:

```cs
public abstract Task BeginBuildAsync(CacheContext context, PluginLoggerBase logger, CancellationToken cancellationToken);

public abstract Task EndBuildAsync(PluginLoggerBase logger, CancellationToken cancellationToken);

public abstract Task<CacheResult> GetCacheResultAsync(BuildRequestData buildRequest, PluginLoggerBase logger, CancellationToken cancellationToken);

public virtual void HandleFileAccess(FileAccessContext fileAccessContext, FileAccessData fileAccessData);

public virtual void HandleProcess(FileAccessContext fileAccessContext, ProcessData processData);

public virtual Task HandleProjectFinishedAsync(FileAccessContext fileAccessContext, BuildResult buildResult, PluginLoggerBase logger, CancellationToken cancellationToken);

> ```

Reviewer 3: This feels superfluous to the link to iplementation provided

Reviewer 3: Also it should probably link the `Deteours` section below (and leave the implementation details info - like the need to disable in-proc node etc. - to be mentioned only in that section. Now some info is duplicated)

Author: Removing the signatures and linked to the `Plugin implementation guidance and simple example design` section.

- Requires `/reportfileaccesses` to light up cache add scenarios.
- The static graph has all the project instances in the same process, makes it easy to find and keep plugin instances in one process.
- MSBuild constructs the static graph and build bottom up, so by the time a project is considered, all of its references and their build results are already present in the Scheduler.

## Enabling from Visual Studio, a temporary workaround

- Ideally, Visual Studio would provide a `ProjectGraph` instance. Until that happens, a workaround is needed.
- The workaround logic activates only when MSBuild detects that it's running under VS.
- When VS evaluates projects via `new Project(..)` (it does this on all the solution projects on solution load), the evaluator will search for and store all detected plugins in a static field on the `BuildManager`.
- The first build request will check the static state for the presence of plugins. If there's a plugin, it will initialize it at that point.
- Plugins will be given the graph entry points instead of the entire graph in this scenario.
- There is currently no way to enable cache add scenarios in Visual Studio.

# Detours (cache add scenario)

> In order for MSBuild to observe the file accesses as part of the build, it uses Detours on the worker nodes. In this way the Scheduler node will events for all file accesses done by the worker nodes. As the Scheduler knows what build request a worker node is working on at any given moment, it is able to properly associate the file access with a build request and dispatch these augmented events to plugins via the plugins' `HandleFileAccess` and `HandleProcess` implementations.

Author: Should be "will emit events"

Note that the Scheduler node cannot use Detours on itself, so the in-proc node is disabled when repoting file accesses. Additionally task yielding is disabled since it would leave to improperly associated file accesses.

## Pipe synchronization

> Because the Detours implementation being used communicates over a pipe, and node communication is also over a pipe, and pipes are async, there is some coordination required for ensuring that file accesses are associated with the proper build request. For example, if a "project finished" signal comes through the node communication pipe, but the detours pipe still has a queue of file accesses which have not been processed yet, those file accesses might be processed after the worker node has moved onto some other project.

Reviewer 3: ```suggestion
Because the Detours implementation being used communicates over a pipe, and nodes communicate over a pipe as well, and pipes are async, there is some coordination required to ensure that file accesses are associated with the proper build request. For example, if a "project finished" signal comes through the node communication pipe, but the detours pipe still has a queue of file accesses which have not been processed yet, those file accesses might be processed after the worker node has moved onto some other project.
```

But I'm not a native speaker though

Author: That's much less awkward sentence structure :)

> To address this problem, when a worker node finishes a project it will emit a dummy file access with a specific format known to MSBuild. When the scheduler not recieves as "project finished" event over the node communication pipe, it will wait to determine that the project is actually finished until it also recieves the dummy file access. This ensures that the all file accesses associated with the project have fully flushed from the pipe before the scheduler deterines the project is finished and schedules new work to the worker node (which would trigger new file accesses).

Author: "scheduler node receives"

I also spelled receives wrong. Twice.

## First build (cache population)

In the very first build there will be no cache hits so the "cache add" scenario will be most relevant here.

For a given project, `GetCacheResultAsync` will be invoked, but will end up returning a cache miss since the cache is empty.

MSBuild will then build the project normally but under a detoured worker node. Because of this, the plugin will recieve `HandleFileAccess` and `HandleProcess` events. In this example implementation we will ignore `HandleProcess`. For `HandleFileAccess`, the plugin will simply store all `FileAccessData`s for a `FileAccessContext` to build up a list of all file accesses during the build. The plugin may decide to avoid storing the entire `FileAccessData` and instead just peel off the data it finds relevant (eg. paths, whether it was a read or write, etc).

> Once MSBuild is done building the project, it will call the plugin's `HandleProjectFinishedAsync`. Now the plugin knows the project is done and can process the results and add them a cache. In general it's only useful to cache successful results, so the plugin should filter out non-success results. The `FileAccessContext` provided can then be used to retrieve the list of `FileAccessData` the plugin recieved. These `FileAccessData` can be processed to understand which files were read and writted as part of the build.

Reviewer 2: Nit: add them **to** a cache

Reviewer 2: Nit; **written**

In our example, we can use the read files to construct a fingerprint for the build request. We can then add the files written during the build ("outputs") to some cache implementation.

The plugin would then create some metadata describing the outputs (eg. the paths and hashes) and the serialized `BuildResult`, and associate it with the fingerprint and put that assocation in the cache.

To illustrate this, consider a project with fingerprint `F` which wrote a single file `O` with hash `H` and had `BuildResult R`. The plugin could create a metadata file `M` which describes the outputs of the build (the path and hash of `O`) as well as the serialized `R`. Using the cache implementation described above, the plugin would write the following two files to the cache:
 - `metadata/F -> M`
 - `content/H -> O`

This can then be used for future builds.

 ## Second Build (cache hits)
 
 In the second build we have a populated cache and so it could be possible to get cache hits.

 For a given project, `GetCacheResultAsync` will be invoked. The plugin can fingerprint the request and use that fingerprint to look up in its cache. If the cache entry exists, it can declare a cache hit.

> In the example above, if all inputs are the same as in the first build, we should end up with a fingerprint `F`. We look up in the metadata part of the cache (file `metadata/F`) and find that it exists. This means we have a cache hit. We can fetch that metadata `M` from the cache and find that is describes the output with path `O` and hash `H`. The plugin would then copy `content/H` to `O` and return the deserialized `BuildResult R` contained in `M` to MSBuild.

Reviewer 2: Nit: **it** describes
